### DISCUSSION TITLE
How to add `libmsquic` to container images

### DISCUSSION THREAD

**Discussion by richlander** [dotnet/dotnet-docker#4823] [General]
Here is a quick POC for adding `libmsquic` to container images. This Dockerfile was created to determine the cost of `libmsquic`. It is about 14MB (as demonstrated by this Dockerfile). Sharing in case it is useful for others.

```Dockerfile
#syntax=docker/dockerfile:1

FROM scratch AS installer
ADD https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb /


FROM mcr.microsoft.com/dotnet/runtime-deps:8.0-preview-jammy
RUN --mount=from=installer,source=/packages-microsoft-prod.deb,target=/packages-microsoft-prod.deb \
<<EOF
set -e
dpkg -i packages-microsoft-prod.deb
apt update
apt install -y libmsquic
apt remove -y packages-microsoft-prod
rm -rf /var/lib/apt/lists/*
EOF
```

**Comment by MichaelSimons** [comment]
What is the advantage of using a multi-stage dockerfile versus downloading the package and installing it in a singe stage?

**Reply by richlander** [reply]
What I did/saw. The second approach has one more layer. Feel free to validate yourself.

```bash
rich@vancouver:~/test$ cat Dockerfile.ubuntu
# syntax=docker/dockerfile:1

FROM scratch AS installer
ADD https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb /


FROM mcr.microsoft.com/dotnet/runtime-deps:8.0-preview-jammy
RUN --mount=from=installer,source=/packages-microsoft-prod.deb,target=/packages-microsoft-prod.deb \
<<EOF
dpkg -i packages-microsoft-prod.deb
apt update
apt install -y libmsquic
apt remove -y packages-microsoft-prod
rm -rf /var/lib/apt/lists/*
EOF
rich@vancouver:~/test$ cat Dockerfile.ubuntu2
# syntax=docker/dockerfile:1

FROM mcr.microsoft.com/dotnet/runtime-deps:8.0-preview-jammy
ADD https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb /

RUN <<EOF
dpkg -i packages-microsoft-prod.deb
apt update
apt install -y libmsquic
apt remove -y packages-microsoft-prod
rm packages-microsoft-prod.deb
rm -rf /var/lib/apt/lists/*
EOF

rich@vancouver:~/test$ docker build -t quic -f Dockerfile.ubuntu .
rich@vancouver:~/test$ docker build -t quic2 -f Dockerfile.ubuntu2 .
rich@vancouver:~/test$ docker inspect quic | grep sha256
        "Id": "sha256:ab4d7a4585570135e31ddcdb12617ad746e32daaff87593bd9d60fc67c544e04",
                "sha256:bce45ce613d34bff6a3404a4c2d56a5f72640f804c3d0bd67e2cf0bf97cb950c",
                "sha256:9859d7cd6a80ff888a44f9d27f2f4876b760ad21b070d11770d7812eac8c9eea",
                "sha256:68cd1df3d5f7a28fe8c69fcef7c85107b21a127af55d53879c44daa28c2885f8",
                "sha256:9df966d347b2b727d7f424bb5128f8149a910c78ecf261a2a4609adbf484a141"
rich@vancouver:~/test$ docker inspect quic2 | grep sha256
        "Id": "sha256:da2cac9cf2ee4121c200eb9dfafa8e3d8c0e5227d63cb3bcbc8fa9bcf0e8db0c",
                "sha256:bce45ce613d34bff6a3404a4c2d56a5f72640f804c3d0bd67e2cf0bf97cb950c",
                "sha256:9859d7cd6a80ff888a44f9d27f2f4876b760ad21b070d11770d7812eac8c9eea",
                "sha256:68cd1df3d5f7a28fe8c69fcef7c85107b21a127af55d53879c44daa28c2885f8",
                "sha256:43f81ad892c4d547bd370ed0300eccbee07e86b1c5fceebee4208fe4b7288b79",
                "sha256:80cc08d6e0c71854bce593af7ad64adcbd99482f56f0d6c78e19d00042f3d592"
```

**Comment by kamranattari4192** [comment]
Hi,
I am trying to add libmsquic in my application I used your instructions in docker file but it does not install libmsquic. At runtime, I am getting **Quic not supported System.PlatformNotSupportedException: System.Net.Quic is not supported on this platform.**
Here is my docker file:

FROM scratch AS installer
ADD https://packages.microsoft.com/config/ubuntu/22.04/packages-microsoft-prod.deb /

FROM mcr.microsoft.com/dotnet/runtime:7.0-jammy AS base
RUN --mount=from=installer,source=/packages-microsoft-prod.deb,target=/packages-microsoft-prod.deb \
<<EOF
dpkg -i packages-microsoft-prod.deb
apt update
apt install -y libmsquic
apt remove -y packages-microsoft-prod
rm -rf /var/lib/apt/lists/*
EOF
WORKDIR /app
EXPOSE 9002/udp
COPY . .
ENTRYPOINT ["dotnet", "MyApp.dll"]

I even tried with your docker file and started the container in interactive mode and tried to find libmsquic with following commands, it does not found it:
root@c541bcd50711:/# ldconfig -p | grep libmsquic
root@c541bcd50711:/# ldd /usr/lib/libmsquic.so.2
ldd: /usr/lib/libmsquic.so.2: No such file or directory
root@c541bcd50711:/# ldd /lib64/libmsquic.so.2
ldd: /lib64/libmsquic.so.2: No such file or directory
root@c541bcd50711:/# ldd /lib/x86_64-linux-gnu/libmsquic.so.2
ldd: /lib/x86_64-linux-gnu/libmsquic.so.2: No such file or directory

**Reply by polarathene** [reply]
You can reference [this example](https://github.com/TechnitiumSoftware/DnsServer/pull/1046/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R17-R18) where I grabbed the package on a plain Ubuntu base image and extracted only the necessary files to copy over to the `chiseled` variant (_half the size of the full `dotnet/aspnet` image_).

If you have trouble, try starting the container for the base image and running each command separately to see if you run into any failures along the way. Your `apt install -y libmsquic` may have failed (_I think I ran into that when I tried `ADD`, only to realize that without `ca-certificates` the package install failed_).

---

With the `dotnet/aspnet` image as base it works fine without curl, so you'd use:

```Dockerfile
# syntax=docker.io/docker/dockerfile:1

FROM mcr.microsoft.com/dotnet/aspnet:8.0
# Add the MS repo to install `libmsquic` to support DNS-over-QUIC:
ADD --link https://packages.microsoft.com/config/debian/12/packages-microsoft-prod.deb /
RUN <<HEREDOC
  dpkg -i packages-microsoft-prod.deb && rm packages-microsoft-prod.deb
  apt update && apt install -y libmsquic
  apt clean -y && rm -rf /var/lib/apt/lists/*
HEREDOC
```

With `ADD --link` you should be able to skip the separate stage + `RUN --mount` as `--link` accomplishes roughly the same result.

**Comment by polarathene** [comment]
> It is about 14MB (as demonstrated by this Dockerfile)

For reference half of that is a `lttng` library which is [only useful when you need logging](https://github.com/microsoft/msquic/issues/1394#issuecomment-808967416) support?

---

For anyone else interested in using the smaller `chisel` base images for runtime, I gave a rather in-depth look into that process here: https://github.com/TechnitiumSoftware/DnsServer/issues/1012#issuecomment-2373082530

For `libmsquic` I checked the two libraries from the package were portable/compatible to [extract and copy over](https://github.com/TechnitiumSoftware/DnsServer/pull/1046/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R17-R18), although the maintainer of the project I contributed that too wasn't too confident in this method being reliable. If that does turn out to be the case, someone could contribute a sliced package to Ubuntu Chisel which might be a better way to resolve that.

The end result was 270MB `dotnet/aspnet` image to the official `dotnet/aspnet` chisel variant at 123MB (_includes `libmsquic` + 6MB project files_).
